compiler/codegen/mli.cpp

```diff
@@ -0,0 +1,912 @@
+
+  gen += this->genServerDispatchSwitch(this->exps);
+  gen += "\n";
+  
+  this->setOutputAndWrite(&this->fiServerBundle, gen);
+
+  return;
+}
+
+void MLIContext::setOutput(fileinfo* fi) {
+  this->info->cfile = fi->fptr;
+  return;
+}
+
+void MLIContext::setOutputAndWrite(fileinfo* fi, const std::string& gen) {
+  this->setOutput(fi);
+  this->write(gen);
+}
+
+void MLIContext::write(const std::string& gen) {
+  fprintf(this->info->cfile, gen.c_str());
+  return;
+}
+
+//
+// We can (as I understand it) cound on Type* being unique across the entire
+// symbol table (IE, you'll never encounter two different Type objects that
+// both end up describing the same concrete type).
+//
+int64_t MLIContext::assignUniqueTypeID(Type* t) {
+  // Prepare a new ID based on the map size (will never overflow).
+  int64_t result = (int64_t) this->typeMap.size();
+
+  if (this->typeMap.find(t) != this->typeMap.end()) {
+    result = this->typeMap[t];
+  } else {
+    this->typeMap[t] = result;
+  }
+
+  return result;
+}
+
+void MLIContext::emitClientWrapper(FnSymbol* fn) {
+  std::string gen;
+
+  this->setOutput(&this->fiClientBundle);
+  fn->codegenHeaderC();
+
+  gen += scope_begin;
+  gen += this->genClientsideRPC(fn);
+  gen += scope_end;
+  gen += "\n";
+
+  this->write(gen);
+  
+  return;
+}
+
+void MLIContext::emitServerWrapper(FnSymbol* fn) {
+  std::string gen;
+
+  // Big long, silly block of manual code generation.
+  gen += this->genComment(toString(fn));
+  gen += "int64_t chpl_mli_swrapper_";
+  gen += this->genFuncNumericID(fn);
+  gen += "(void)";
+  gen += scope_begin;
+
+  gen += this->genServersideRPC(fn);
+  gen += "return 0;\n";
+ 
+  gen += scope_end;
+  gen += "\n";
+
+  this->setOutputAndWrite(&this->fiServerBundle, gen);
+ 
+  return;
+}
+
+std::string MLIContext::genDebugPrintCall(FnSymbol* fn, const char* pfx) {
+  std::string msg;
+
+  msg += "Calling: ";
+  msg += toString(fn);
+
+  return this->genDebugPrintCall(msg.c_str(), pfx);
+}
+
+std::string MLIContext::genDebugPrintCall(const char* msg, const char* pfx) {
+  std::string gen;
+
+  gen += "printf(\"";
+
+  if (pfx && pfx != "") {
+    gen += pfx;
+    gen += " ";
+  }
+
+  gen += "%%s\\n\", \"";
+  gen += msg;
+  gen += "\");\n";
+
+  return gen;
+}
+
+std::string MLIContext::genFuncNumericID(FnSymbol* fn) {
+  return std::to_string((int64_t) fn->id);
+}
+
+std::string MLIContext::genServerWrapperCall(FnSymbol* fn) {
+  std::string gen;
+
+  gen += "chpl_mli_swrapper_";
+  gen += this->genFuncNumericID(fn);
+  gen += "();\n";
+
+  return gen;
+}
+  
+std::string
+MLIContext::genServerDispatchSwitch(const std::vector<FnSymbol*>& fns) {
+  std::string gen;
+
+  gen += "int64_t chpl_mli_sdispatch";
+  gen += "(int64_t function)";
+  gen += scope_begin;
+  gen += this->genNewDecl("int", "err");
+  gen += "switch (function)";
+  gen += scope_begin;
+
+  for_vector(FnSymbol, fn, fns) {
+    if (not fn->hasFlag(FLAG_EXPORT) || fn->hasFlag(FLAG_GEN_MAIN_FUNC)) {
```

Comment:
Don't you already have a function that computes this which you could use?

Reply:
Also, this does not take into account the MOD_INTERNAL stuff that the other calls did, is that intentional?

Reply:
I believe I do, and just missed that case. MOD_INTERNAL was just to skip internal modules.

Reply:
Ah, they already won't be included because the earlier MOD_INTERNAL check prevented them from being added to fns?

Reply:
Internal modules are just skipped entirely, yes.